### PR TITLE
Lower dimensional datasets

### PR DESCRIPTION
**Pull request by reguly**
This introduces support in code generators and backends for lower dimensional datasets. These are even further simplified examples that Jammy's bug pointed out.

Current issues:
- 1D and 2D datasets not written to HDF5 during ops_dump_to_hdf5
- MPI not supported yet. Need special MPI transfers that scatter data written on a subset of processes to other processes

### CODE REVIEW

**Comment by spjammy** [comment]
The tests reproduce the error we are facing currently.

**Comment by gihanmudalige** [comment]
Testing the lowdim application - OpenCL on the CPU core dumps. 
octon {feature/lowdim} ~/OPS-GIT/OPS/apps/c/lowdim_test$ ./lowdim_opencl OPS_CL_DEVICE=0 

OPS_cl_device = 0 
Number of OpenCL platforms = 2 
Number of devices on platform 1 = 1
OpenCL Running on platform 1 device 0 : Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz
Kernel source file ./OpenCL/set_val.cl succesfuly read.
Compiling set_val 0 source -- start 
compiling set_val -- done
Kernel source file ./OpenCL/calc.cl succesfuly read.
Compiling calc 0 source -- start 
Segmentation fault (core dumped)

**Comment by gihanmudalige** [comment]
All the original tests pass. But the low_dim tests produces the following for opencl :

octon {feature/lowdim} ~/OPS-GIT/OPS/apps/c/lowdim_test$ ./lowdim_opencl OPS_CL_DEVICE=0 OPS_BLOCK_SIZE_X=512 OPS_BLOCK_SIZE_Y=1 
 OPS_block_size_x = 512 

 OPS_block_size_y = 1 

 OPS_cl_device = 0 
Number of OpenCL platforms = 2 
Number of devices on platform 1 = 1
OpenCL Running on platform 1 device 0 : Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz
Kernel source file ./OpenCL/set_val.cl succesfuly read.
Compiling set_val 0 source -- start 
compiling set_val -- done
Kernel source file ./OpenCL/calc.cl succesfuly read.
Compiling calc 0 source -- start 
compiling calc -- done
*** Error in `./lowdim_opencl': free(): invalid next size (normal): 0x0000000003b2e3f0 ***
Aborted (core dumped)


Basically the same error as before.

**Comment by gihanmudalige** [comment]
have rebased this now on current master -- however still the opencl issue persists.

The opencl version on GPUs run fine. 

But the CPU version does not. It almost seems like an Intel opencl problem. But then all other apps work fine with Intel OpenCL.  Could there be something special in the low_dim code that generates opencl that does not work on CPUs ?

**Comment by gihanmudalige** [comment]
I had a go at debugging this today.  gdb points to the following line:
ret = clBuildProgram(OPS_opencl_core.program, 1, &OPS_opencl_core.device_id, buildOpts, NULL, NULL);
Have tried using the offline opencl .cl file compiler and that also builds the kernels fine. So I have no idea why this is  breaking.

**Comment by reguly** [comment]
@gihanmudalige as discussed this is not a bug introduced by this branch. Adding an issue, please approve if all other tests complete

**Comment by reguly** [comment]
It seems that the xdim, ydim variables are still 0 in OpenCL. Investigating a bit it seems OpenCL does not support setting global scope variables. Therefore we need to pass these extents as function arguments. I suggest we have an array filled with these values as we do in CUDA. Of course they will need to be passed to the suer kernel as well.

Perhaps we should delay fixing this until we can pull in the accessors feature, then it will be much easier to do, because we won't need to pass them to the user function.

**Comment by gihanmudalige** [comment]
Yes. I had a play around with this .. and did add the extents as extra arguments. .. but could not get it to copy as an array (yet). 
Ok, perhaps we can wait for the accessor features.

**Comment by reguly** [comment]
Ok - if things are fine otherwise, please approve.
